File: src/modules/mangaworld.ts
import { SearchResult, Manga, Chapter } from '../classes/interfaces'
import Module from '../classes/Module'
import axios from 'axios'
import cheerio from 'cheerio'

class Mangaworld implements Module {
  id = 'mw'
  name = 'MangaWorld'

  search(query: string) {
    return new Promise(async (resolve: (value: SearchResult[]) => void) => {

      const $ = cheerio.load((await axios.get(`https://www.mangaworld.io/archive?keyword=${query}`)).data)

      let results: SearchResult[] = []
      $('.entry').each((i, el) => {

        const mid = $('a', el).attr('href')?.split('/manga/')[1]+''
        const title = $('a', el).attr('title')+''

        results.push({id: mid, title: title})
      })

      resolve(results)

    })
  }

  manga(mangaid: string) {
    return new Promise(async (resolve: (value: Manga) => void) => {

      let chapters: Chapter[] = []

      let data = {
        synopsis: '',
        author: '',
        artist: '',
        img: '',
        chapters: chapters,
        sourceurl: 'https://www.mangaworld.io/manga/'+mangaid
      }

      const $ = cheerio.load((await axios.get(`https://www.mangaworld.io/manga/${mangaid}`)).data)

      data.synopsis = $('#noidungm').text()

      $(".col-12.col-md-6", $('.meta-data')).each((i, div) => {
        if($("span", div).text().includes('Autor')) {
          data.author = $("a", div).text()
        }
        if($("span", div).text().includes("Artist")) {
          data.artist = $("a", div).text()
        }
      })

      data.img = $("img.rounded").attr("src")+'';

      $('.chapter').each((i, entry) => {
        let cap = {
          title: '',
          id: '',
          date: new Date()
        }

        const chap = $('.chap', entry)

        cap.title = $('span', chap).text()
        cap.id = chap.attr('href')?.split('/read/')[1]+''
        cap.date = psDate($('i', chap).text())

        data.chapters.push(cap)
      })

      resolve(data)
    })
  }

  chapter(manga: string, id: string) {
    return new Promise(async (resolve: (value: string[]) => void) => {

      let result: string[] = []
      
      const body = (await axios.get(`https://www.mangaworld.io/manga/${manga}/read/${id}`)).data
      const json = JSON.parse(body.split('$MC=(window.$MC||[]).concat(')[1].split(')</script>')[0]);
      const pages: string[] = json.o.w[0][2].chapter.pages
      const $ = cheerio.load(body)

      const firstimage = $('#page').find('img').attr('src')?.split('/')
      firstimage?.pop()
      const baseurl = firstimage?.join('/')

      pages.forEach(page => {
        result.push(baseurl+'/'+page)
      })

      resolve(result)

    })
  }
}

export default new Mangaworld()

function psDate(input: string) {
  let mese = 0;
  let parts = input.split(' ');

  switch (parts[1]) {
      case "Gennaio":
          mese = 0;
          break;
  
      case "Febbraio":
          mese = 1;
          break;

      case "Marzo":
          mese = 2;
          break;

      case "Aprile":
          mese = 3;
          break;

      case "Maggio":
          mese = 4;
          break;
  
      case "Giugno":
          mese = 5;
          break;

      case "Luglio":
          mese = 6;
          break;

      case "Agosto":
          mese = 7;
          break;

      case "Settembre":
          mese = 8;
          break;
  
      case "Ottoble":
          mese = 9;
          break;

      case "Novembre":
          mese = 10;
          break;

      case "Dicembre":
          mese = 11;
          break;
  }

  let date = new Date()
  date.setFullYear(parseInt(parts[2]))
  date.setMonth(mese)
  date.setDate(parseInt(parts[0]))

  return date
}

File: pkg/modules/mangaworld/mangaworld.go
package mangaworld

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/Minettyx/FoolslideProxy/pkg/types"
	"github.com/Minettyx/FoolslideProxy/pkg/utils"
	"github.com/dop251/goja"

	"github.com/PuerkitoBio/goquery"
)

//go:embed aes.min.js
var aesMinJs string

type mangaWorld struct {
	baseUrl string
	name    string
	id      string
}

var Module = mangaWorld{
	baseUrl: "https://www.mangaworld.ac/",
	name:    "MangaWorld",
	id:      "mw",
}

var ModuleAdult = mangaWorld{
	baseUrl: "https://www.mangaworldadult.net/",
	name:    "MangaWorld adult",
	id:      "mwa",
}

var _ types.Module = mangaWorld{}

func (c mangaWorld) Id() string {
	return c.id
}
func (c mangaWorld) Name() string {
	return c.name
}
func (c mangaWorld) Flags() types.ModuleFlags {
	return types.ModuleFlags{}
}

func client(req *http.Request) (*http.Response, error) {

	hclient := http.Client{
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
	res, err := hclient.Do(req)
	if err != nil {
		return nil, err
	}

	bodyB, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	body := string(bodyB)
	body, _ = utils.StrBetweenFirst(body, "<script>", "location.href")
	body = strings.Replace(body, "document.cookie", "const finalcookie", 1)

	code := aesMinJs + "\n\n" + body

	vm := goja.New()
	_, err = vm.RunString(code)
	if err != nil {
		return nil, err
	}

	var finalcookie string
	err = vm.ExportTo(vm.Get("finalcookie"), &finalcookie)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Cookie", finalcookie)

	finalres, err := hclient.Do(req)
	if err != nil {
		return finalres, err
	}

	if finalres.StatusCode == 301 || finalres.StatusCode == 302 {
		loc, _ := finalres.Location()
		recreq, _ := http.NewRequest("GET", loc.String(), nil)
		return client(recreq)
	}

	return finalres, err
}

func (c mangaWorld) Search(query string) ([]types.SearchResult, error) {
	req, _ := http.NewRequest("GET", c.baseUrl+"archive?keyword="+query, nil)
	res, err := client(req)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	results := []types.SearchResult{}

	doc.Find(".entry").Each(func(i int, s *goquery.Selection) {
		mid, _ := s.Find("a").Attr("href")
		pts := strings.Split(mid, "/manga/")
		if len(pts) < 2 {
			return
		}
		mid = pts[1]

		title := s.Find(".manga-title").Text()

		image, _ := s.Find("img").Attr("src")

		results = append(results, types.SearchResult{
			Id:    mid,
			Title: title,
			Image: image,
		})
	})

	return results, nil
}

func (c mangaWorld) Manga(id string) (*types.Manga, error) {
	req, _ := http.NewRequest("GET", c.baseUrl+"manga/"+id, nil)
	res, err := client(req)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	synopsis := doc.Find("#noidungm").Text()
	author := ""
	artist := ""

	doc.Find("div.col-12.col-md-6").Each(func(i int, s *goquery.Selection) {
		inx := s.Find("span").Text()
		val := s.Find("a").Text()

		if inx == "Autore: " {
			author = val
		} else if inx == "Artista: " {
			artist = val
		}
	})

	img, _ := doc.Find(".rounded").Attr("src")

	chapters := []types.Chapter{}

	doc.Find(".chapter").Each(func(i int, s *goquery.Selection) {
		chtitle := s.Find("span").Text()

		chid, _ := s.Find("a").Attr("href")
		pts := strings.Split(chid, "/read/")
		if len(pts) < 2 {
			return
		}
		chid = pts[1]

		datestr := s.Find("i").Text()
		dat, err := parseDate(datestr)
		if err != nil {
			return
		}

		chapters = append(chapters, types.Chapter{
			Title: chtitle,
			Id:    chid,
			Date:  *dat,
		})
	})

	manga := types.Manga{
		Synopsis:  synopsis,
		Author:    author,
		Artist:    artist,
		Img:       img,
		Chapters:  chapters,
		Sourceurl: c.baseUrl + "manga/" + id,
	}

	return &manga, nil
}

func (c mangaWorld) Chapter(manga, id string) ([]string, error) {
	req, _ := http.NewRequest("GET", c.baseUrl+"manga/"+manga+"/read/"+id, nil)
	res, err := client(req)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	jsonstr, err := utils.StrBetweenFirst(string(body), `"pages":`, "]")
	jsonstr += "]"
	if err != nil {
		return nil, err
	}

	var pages []string
	err = json.Unmarshal([]byte(jsonstr), &pages)
	if err != nil {
		return nil, err
	}

	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(body))
	if err != nil {
		return nil, err
	}

	firstimage, _ := doc.Find("#page").Find("img").Attr("src")
	pts := strings.Split(firstimage, "/")
	pts = pts[0 : len(pts)-1]
	baseimageurl := strings.Join(pts, "/")

	for i := range pages {
		pages[i] = baseimageurl + "/" + pages[i]
	}

	return pages, nil
}

func (c mangaWorld) Latest() ([]types.LatestResult, error) {
	return []types.LatestResult{}, nil
}

func (c mangaWorld) Popular() ([]types.PopularResult, error) {
	return []types.PopularResult{}, nil
}

func parseDate(input string) (*time.Time, error) {
	var mese time.Month
	parts := strings.Split(input, " ")

	if len(parts) < 3 {
		return nil, fmt.Errorf("Error parsing date string")
	}

	switch parts[1] {
	case "Gennaio":
		mese = time.January
		break

	case "Febbraio":
		mese = time.February
		break

	case "Marzo":
		mese = time.March
		break

	case "Aprile":
		mese = time.April
		break

	case "Maggio":
		mese = time.May
		break

	case "Giugno":
		mese = time.June
		break

	case "Luglio":
		mese = time.July
		break

	case "Agosto":
		mese = time.August
		break

	case "Settembre":
		mese = time.September
		break

	case "Ottobre":
		mese = time.October
		break

	case "Novembre":
		mese = time.November
		break

	case "Dicembre":
		mese = time.December
		break
	}

	year, err := strconv.Atoi(parts[2])
	if err != nil {
		return nil, err
	}

	day, err := strconv.Atoi(parts[0])
	if err != nil {
		return nil, err
	}

	dat := time.Date(year, mese, day, 0, 0, 0, 0, time.UTC)
	return &dat, nil
}
